docs/proposals/THV-2106-virtual-mcp-server.md

```diff
@@ -0,0 +1,1035 @@
+
+### Two Authentication Boundaries
+
+Virtual MCP operates with two distinct, independent authentication boundaries:
+
+1. **Incoming Authentication** (Client → Virtual MCP)
+   - **Purpose**: Validates client requests to Virtual MCP
+   - **Token Audience**: `aud=vmcp`
+   - **Configuration**: `incomingAuth` (K8s) / `incoming_auth` (CLI)
+   - **Implementation**: Uses existing ToolHive auth middleware (OIDC, local, anonymous)
+
+2. **Outgoing Authentication** (Virtual MCP → Backends)
+   - **Purpose**: Per-backend token exchange for accessing backend services
+   - **Token Audience**: Backend-specific (e.g., `aud=github-api`, `aud=jira-api`)
+   - **Configuration**: `outgoingAuth` (K8s) / `outgoing_auth` (CLI)
+   - **Implementation**: RFC 8693 token exchange (references THV-2063)
+
+```mermaid
+graph TB
+    Client[Client<br/>Token: aud=vmcp]
+
+    subgraph "Auth Boundary #1: Client → Virtual MCP"
+        VMCPAuth[Virtual MCP<br/>Validates: aud=vmcp<br/>Enforces authz policies]
+    end
+
+    subgraph "Auth Boundary #2: Virtual MCP → Backends"
+        TokenEx1[Token Exchange<br/>subject=vmcp-token<br/>audience=github-api]
+        TokenEx2[Token Exchange<br/>subject=vmcp-token<br/>audience=jira-api]
+    end
+
+    subgraph "Backend Validation"
+        GitHub[GitHub MCP<br/>Validates: aud=github-api]
+        Jira[Jira MCP<br/>Validates: aud=jira-api]
+    end
+
+    Client --> VMCPAuth
+    VMCPAuth --> TokenEx1
+    VMCPAuth --> TokenEx2
+    TokenEx1 --> GitHub
+    TokenEx2 --> Jira
+```
+
+### Zero-Trust Security Model
+
+Virtual MCP adheres to a zero-trust security architecture where compromising one layer does not automatically grant access to others:
+
+**Security Properties:**
+
+1. **Backend Authentication Independence**: Each backend MCP server MUST enforce its own authentication and authorization, regardless of access method (direct or via Virtual MCP). This is a deployment requirement, not optional.
+
+2. **Token Exchange Per Backend**: Virtual MCP performs RFC 8693 token exchange independently for each backend, requesting tokens with backend-specific audiences. Each exchanged token is scoped only to its target backend.
+
+3. **Defense in Depth**: Even if Virtual MCP is compromised:
+   - Attacker must possess a valid user token (with `aud=vmcp`) to perform exchanges
+   - Backend servers independently validate audience, expiry, and signatures
+   - Each backend receives only tokens scoped for its specific audience
+   - Audit logs maintained at both Virtual MCP and backend levels
+
+4. **Credential Isolation**: Virtual MCP does not store long-lived credentials for backend APIs. It only stores OAuth client credentials for token exchange, which require a valid user token to be useful.
+
+5. **Network Isolation** (Recommended): Backend MCP servers SHOULD NOT be directly accessible outside their security boundary (e.g., not exposed via public LoadBalancer). Access SHOULD be routed through Virtual MCP or authenticated ingress.
```

Comment:
Although it's technically bypassable if users wanted to, but can we do extra checks on the proxy too? Similar to the way service mesh sidecars work with verifying requests?

Reply:
Nope nothing in mind - was just spitballing ideas. I'd have to look into the way service meshes work to have more input on a solution to this 👍